### DISCUSSION TITLE
What is test in GitOps?

### DISCUSSION THREAD

**Discussion by schlomo** [open-gitops/project#87] [Q&A] [answered]
I have a conceptional question about the CI loop: It mentions test but deployment happens in the CD loop, so do you mean only unit tests here? Or what sort of tests to we expect there? How would a test-and-deploy-in-production setup look like?

![image](https://user-images.githubusercontent.com/101384/107774310-1f464c00-6d3f-11eb-9b5c-4a16042c5d1b.png)

Slide grabbed from @cdavisafc, Source: https://fosdem.org/2021/schedule/event/gitops_working_group/

**Comment by cdavisafc** [comment, marked as answer]
Yes, Schlomo, that is exactly what I intended. The GitOps loop on the right does a deployment, perhaps into an integration environment (say, a K8s cluster) and then monitor-manage-commitToGit closes the OODA loop in that env. You can also then imagine another uber-workflow that issues a PR against a config repo for a production environment so there would be another GitOps OODA loop for that prod environment.

**Reply by lloydchang** [reply]
Hi @cdavisafc @schlomo 

My understanding is beyond Unit Testing, additional possibilities during the Continuous Integration (CI) loop are Contract Testing and Security Testing, such as:
• Integration Contract Testing
• Consumer-Driven Contract Testing (CDC Testing)
• Static Application Security Testing (SAST)
• Open Source Dependency Scans
• Container Image Scans

Some examples with details:
• https://www.infosys.com/services/it-services/documents/microservices-testing-strategies.pdf by Arvind Sundar at @Infosys
• https://www.mock-server.com/#why-use-mockserver by @jamesdbloom
• https://martinfowler.com/articles/practical-test-pyramid.html#ContractTests by @hamvocke
• https://docs.pact.io/#consumer-driven-contracts by @mefellows at @pact-foundation
• https://owasp.org/www-community/Source_Code_Analysis_Tools by @OWASP
• https://www.atlassian.com/trust/security/vulnerability-management by @atlassian

Thank you!
